sycl/include/CL/sycl/detail/kernel_impl.hpp

```diff
@@ -29,118 +28,123 @@ class program_impl;
 
 class kernel_impl {
 public:
+  /// Constructs a SYCL kernel instance from a PiKernel
+  ///
+  /// @param Kernel is a valid PiKernel instance
+  /// @param SyclContext is a valid SYCL context
   kernel_impl(RT::PiKernel Kernel, const context &SyclContext);
 
+  /// Constructs a SYCL kernel instance from a SYCL program
+  ///
+  /// @param Kernel is a valid PiKernel instance
+  /// @param SyclContext is a valid SYCL context
+  /// @param ProgramImpl is a valid instance of program_impl
+  /// @param IsCreatedFromSource is a flag that indicates whether program
+  /// is created from source code
   kernel_impl(RT::PiKernel Kernel, const context &SyclContext,
               std::shared_ptr<program_impl> ProgramImpl,
-              bool IsCreatedFromSource)
-      : Kernel(Kernel), Context(SyclContext), ProgramImpl(ProgramImpl),
-        IsCreatedFromSource(IsCreatedFromSource) {
-
-    RT::PiContext Context = nullptr;
-    PI_CALL(RT::piKernelGetInfo, Kernel, CL_KERNEL_CONTEXT, sizeof(Context),
-            &Context, nullptr);
-    auto ContextImpl = detail::getSyclObjImpl(SyclContext);
-    if (ContextImpl->getHandleRef() != Context)
-      throw cl::sycl::invalid_parameter_error(
-          "Input context must be the same as the context of cl_kernel");
-    PI_CALL(RT::piKernelRetain, Kernel);
-  }
+              bool IsCreatedFromSource);
 
-  // Host kernel constructor
+  /// Constructs a SYCL kernel for host device
+  ///
+  /// @param SyclContext is a valid SYCL context
+  /// @param ProgramImpl is a valid instance of program_impl
   kernel_impl(const context &SyclContext,
-              std::shared_ptr<program_impl> ProgramImpl)
-      : Context(SyclContext), ProgramImpl(ProgramImpl) {}
-
-  ~kernel_impl() {
-    // TODO catch an exception and put it to list of asynchronous exceptions
-    if (!is_host()) {
-      PI_CALL(RT::piKernelRelease, Kernel);
-    }
-  }
+              std::shared_ptr<program_impl> ProgramImpl);
 
+  ~kernel_impl();
+
+  /// Gets a valid OpenCL interoperability kernel
+  ///
+  /// The requirements for this method are described in section 4.3.1
+  /// of the SYCL specification.
+  ///
+  /// @return a valid cl_kernel instance
   cl_kernel get() const {
-    if (is_host()) {
+    if (is_host())
       throw invalid_object_error("This instance of kernel is a host instance");
-    }
-    PI_CALL(RT::piKernelRetain, Kernel);
-    return pi::cast<cl_kernel>(Kernel);
-  }
 
-  bool is_host() const { return Context.is_host(); }
-
-  context get_context() const { return Context; }
+    PI_CALL(RT::piKernelRetain, MKernel);
+    return pi::cast<cl_kernel>(MKernel);
+  }
 
+  /// Check if this kernel is a SYCL host device kernel
+  ///
+  /// @return true if a kernel is a valid SYCL host device kernel
+  bool is_host() const { return MContext.is_host(); }
+
+  /// Get the context that this kernel is defined for.
+  ///
+  /// The value returned must be equal to that returned by
+  /// get_info<info::kernel::context>().
```

Comment:
Do we really need this method?
Can we replace all usages of this method with `get_info` call remove this method?

NOTE: I understand that we can't do the same for the `kernel` class, but I think it can be done for `kernel_impl`.
NOTE2: the same should be applied to get_program.